### ISSUE TITLE
Basic theory of this program explained in the README

### ISSUE THREAD

**Issue by daviddahl** [Podcastindex-org/aggregator#2]
Hi.

I am poking through the podcastindex code repos and I find this program interesting. Can you add to the README who would run this, what this program does and why?

Sorry to ask such a dumb question, I just think it would spur more interest for developers to help out if the vision was better articulated.

Perhaps also you could add a few notes on what the roadmap might look like?

**Comment by daveajones** [comment]
That’s a good idea. :-)

It’s an application that takes a list of podcast feeds from a database, attempts to download them to a file and then updates the database to reflect that they are ready for parsing by the parser component which hasn’t been released yet. 

I’ll update the README to better reflect that.

**Comment by daviddahl** [comment]
I'm curious, why does this program need to copy the data to a file? Could you not have a trigger & procedure that validates each feed - and updates the database that the feed is ready for parsing? I am wondering if this application would not benefit from a message queue that is an event-driven model to validate feeds and notify or publish as each feed is created / updated / deleted.

**Comment by daveajones** [comment]
This is one part of a pair of scripts that pull and parse the podcast feeds.  This program grabs a list of feeds from the DB and pulls them down into files and then updates the database that they are ready for parsing.  Then the parser comes through and reads the files and parses their contents into the DB.  

I'm not sure I follow you with:  "Could you not have a trigger & procedure that validates each feed - and updates the database that the feed is ready for parsing?"  Could you elaborate?

**Comment by daviddahl** [comment]
Ah, Ok, I see - I did not understand but have re-read some of the code. 

It seems like you are doing this synchronously:
1. Query database for (some) feeds
2. iterate through all 1 at a time: https://github.com/Podcastindex-org/aggregator/blob/master/aggrivate/aggrivate.js#L169
3. Get the rss data from the server - if possible
4. write the data to disk 
5. Parse feed
6. Update database

You can do this in parallel with forked processes in node (via its `cluster` functionality) quite easily but would need to (ideally) add a message queue for the workers to work from.

See: https://node.readthedocs.io/en/latest/api/cluster/

And a simple message queue is https://github.com/smrchy/rsmq

In this design, on start/on interval, you would 

1. Query the db, add each primary key of each rss feed (to fetch) to a message queue in redis
2. Fork a group of workers to pop the IDs off of the queue
3. The worker gets the db record, performs the fetch
4. The results of the fetch are stored on disk / in redis, the forked child dies off
5. Storage of the rss data in another queue triggers the parent process to fork another type of child that parses/validates the single feed data
6. The original record is updated in the database and the forked child dies off

This would allow you to break up the code into small classes that do one (more simple) thing each, and the workers would make all of this event driven and parallelizable - faster and scalable. 

Anyway, that is how I would do it - it would also allow you to write a set of simple test cases. 

Naturally, to drive adoption, you would want to publish a `docker-compose.yml` for interested operators.

**Comment by daveajones** [comment]
Thank you for such a clear and thorough explanation of your idea. I can definitely see a queue based flow being better in many ways.  

To be clear, you’re talking about forking a new aggregator and/or parser node process for each feed as it comes up in the queue instead of running everything asynchronously inside the same process?

In that set up it would be a queue, a pull fetcher, a parse fetcher and whatever number of children running correct?

**Comment by daviddahl** [comment]
> In that set up it would be a queue, a pull fetcher, a parse fetcher and whatever number of children running correct?

Exactly - and you can have a less complex flow control and easier to test set of modules. Also, your code is executed in a forked process and then is cleaned up - so your main daemon is the only long-lived process. 

Happy to help if you want to try this design out.

**Comment by daveajones** [comment]
Yes, I do want to head this direction.  I think it would allow for greater density on the aggregator nodes.  I may ping you for advice when I get to that point?

**Comment by daviddahl** [comment]
> Yes, I do want to head this direction.  I think it would allow for greater density on the aggregator nodes.  I may ping you for advice when I get to that point?

Absolutely. I'm interested in helping as well

**Comment by sarastinishi** [comment]
[Pull request](https://github.com/Podcastindex-org/aggregator/pull/3) any thoughts @daveajones ?
